test/memory-store-test.ts

```diff
@@ -0,0 +1,111 @@
+
+import { Options } from '../dist/esm/index.js'
+import MemoryStore from '../dist/esm/memory-store.js'
+
+describe('memory store test', () => {
+	it('sets the value to 1 on first call to `increment`', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(1)
+	})
+
+	it('increments the key for the store when `increment` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		await store.increment(key)
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(2)
+	})
+
+	it('resets the count for a key in the store when `resetKey` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		await store.increment(key)
+		await store.resetKey(key)
+
+		const { totalHits } = await store.increment(key)
+		expect(totalHits).toEqual(1)
+	})
+
+	it('resets the count for all keys in the store when `resetAll` is called', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const keyOne = 'test-store-one'
+		const keyTwo = 'test-store-two'
+
+		await store.increment(keyOne)
+		await store.increment(keyTwo)
+		await store.resetAll()
+
+		const { totalHits: totalHitsOne } = await store.increment(keyOne)
+		const { totalHits: totalHitsTwo } = await store.increment(keyTwo)
+		expect(totalHitsOne).toEqual(1)
+		expect(totalHitsTwo).toEqual(1)
+	})
+
+	it('resets the count for all the keys in the store when the timeout is reached', async () => {
+		const store = new MemoryStore()
+		store.init({ windowMs: 50 } as Options)
+		const keyOne = 'test-store-one'
+		const keyTwo = 'test-store-two'
+
+		await store.increment(keyOne)
+		await store.increment(keyTwo)
+
+		setTimeout(async () => {
+			const { totalHits: totalHitsOne } = await store.increment(keyOne)
+			const { totalHits: totalHitsTwo } = await store.increment(keyTwo)
+
+			expect(totalHitsOne).toEqual(1)
+			expect(totalHitsTwo).toEqual(1)
+		}, 60)
+	})
+
+	it('can run in electron where setInterval does not return a Timeout object with an unset function', async () => {
+		const originalSetInterval = setInterval
+		let timeoutId = 1
+		let realTimeoutId: NodeJS.Timer
+		// @ts-expect-error We want to not return a `Timer` instance for testing
+		jest.spyOn(global, 'setTimeout').mockImplementation((callback, timeout) => {
+			realTimeoutId = originalSetInterval(callback, timeout)
+			return timeoutId++
+		})
+
+		const store = new MemoryStore()
+		store.init({ windowMs: -1 } as Options)
+		const key = 'test-store'
+
+		try {
+			const { totalHits } = await store.increment(key)
+			expect(totalHits).toEqual(1)
+		} finally {
+			// @ts-expect-error `realTimeoutId` is already set in the `spyOn` call
+			clearTimeout(realTimeoutId)
+		}
+	})
```

Comment:
Does Jest automatically undo `spyOn` mocks at the end of the test? I think you needed to call `jest.restoreAllMocks()` - but it should ideally be outside of the test, in an after* method so that it will run even if the test fails. I suppose in this case we could just do `const spy = jest.spyOn(global, 'setTimeout')...` and then do `spy.mockRestore()` in the `finally` block, although that makes me a bit nervous that something could go wrong outside of the `try`.

Also, having that `clearTimeout` inside the test also makes me nervous, although, come to think of it, none of the other tests even attempt to clean up after their memory stores...

Reply:
Good catch, we should be calling `restoreAllMocks` in the `finally` block.

Reply:
> none of the other tests even attempt to clean up after their memory stores

How would we go about that?

Reply:
> How would we go about that?

That's a good question... there's not really a good way to do it right now, although I suppose using fake timers for every test and then resetting things to real timers would do the trick.